cbmc/proofs/ParseDNSReply/ParseDNSReply_harness.c

```diff
@@ -0,0 +1,165 @@
+// Given unconstrained values in harness
+uint8_t *buffer;
+size_t buffer_size;
+
+// Abstraction of prvReadNameField
+uint8_t *prvReadNameField(uint8_t *pucByte,
+			  size_t xSourceLen,
+			  char *pcName,
+			  size_t xDestLen){
+
+  // Preconditions
+
+  // pointers are valid pointers into buffers
+  __CPROVER_assert(xSourceLen == 0 ||
+		   (buffer <= pucByte && pucByte < buffer + buffer_size),
+		   "pucByte valid pointer");
+  //  __CPROVER_assume(name <= pcName && pcName < name + name_size);
+  (void) *(pcName); // can only test for valid pointer
+
+  // lengths are valid values for space remaining in the buffers
+  __CPROVER_assert(pucByte + xSourceLen <= buffer + buffer_size,
+		   "xSourceLen valid length");
+  //  __CPROVER_assume(pcName + xDestLen <= name + name_size);
+  (void) *(pcName + (xDestLen-1)); // can only test for valid pointer
+
+  // CBMC loop unwinding: bounds depend on xSourceLen and xDestLen
+  __CPROVER_assert(xSourceLen <= NETWORK_BUFFER_SIZE,
+		   "xSourceLen loop unwinding");
+  //  __CPROVER_assume(xDestLen <= NAME_SIZE);
+
+  // Buffer overflow from integer overflow in expression len < xDestLen-1
+  // xDestLen == 254 in code
+  __CPROVER_assert(xDestLen > 0, "xDestLen nonzero");
+
+  SAVE_OLD(pucByte, uint8_t *);
+  SAVE_OLD(pcName, char *);
+  SAVE_OLD(xSourceLen, size_t);
+  SAVE_OLD(xDestLen, size_t);
+
+  // Function body
+  char bit;
+  size_t offset;
+  uint8_t *rc = bit ? pucByte + offset : 0;
+  __CPROVER_assum(offset <= NETWORK_BUFFER_SIZE);
+
+  // Postconditions
+  __CPROVER_assume((rc == 0) ||
+		   (rc - OLD(pucByte) >= 1 &&
+		    rc - OLD(pucByte) <= OLD(xSourceLen) &&
+		    rc - OLD(pucByte) <= OLD(xDestLen)+2 &&
+		    pucByte == OLD(pucByte) &&
+		    pcName == OLD(pcName) &&
+		    buffer <= rc && rc <= buffer + buffer_size));
+
+  return rc;
+}
+
+// Abstraction of prvSkipNameField
+uint8_t *prvSkipNameField( uint8_t *pucByte, size_t xSourceLen ){
+
+  // Preconditions
+
+  // pointer is valid pointer into buffer
+  __CPROVER_assert(xSourceLen == 0 ||
+		   (buffer <= pucByte && pucByte < buffer + buffer_size),
+		   "pucByte valid pointer");
+
+  // length is valid value for space remaining in the buffer
+  __CPROVER_assert(pucByte + xSourceLen <= buffer + buffer_size,
+		   "xSourceLen valid length");
+
+  // CBMC loop unwinding: bound depend on xSourceLen
+  __CPROVER_assert(xSourceLen <= NETWORK_BUFFER_SIZE,
+		   "xSourceLen loop unwinding");
+
+  SAVE_OLD(pucByte, uint8_t *);
+  SAVE_OLD(xSourceLen, size_t);
+
+  // Function body
+  char bit;
+  size_t offset;
+  uint8_t *rc = bit ? pucByte + offset : 0;
+  __CPROVER_assum(offset <= NETWORK_BUFFER_SIZE);
+
+  // Postconditions
+  __CPROVER_assume((rc == 0) ||
+		   (rc - OLD(pucByte) >= 1 &&
+		    rc - OLD(pucByte) <= OLD(xSourceLen) &&
+		    pucByte == OLD(pucByte) &&
+		    buffer <= rc && rc <= buffer + buffer_size));
+
+  return rc;
+
+}
+
+void harness() {
```

Comment:
Can you please ensure that all of the functions have comment headers? Target audience is an embedded developer that is not familiar with CBMC.